External Method Tables

Keno opened the pull request:
This PR implements a way to keep tables of methods that are
not part of the internal method table, but still participate
in the special support we have for keeping tables of methods,
in particular unification through precompilation and efficient
lookup. The intended design use case is to allow for method overlay
tables for various non-CPU backends (e.g. GPU and TPU). These
backends would like to modify basic function like `sin` to
perform better on the device in question (or in the case of TPU
to define them over non-LLVM intrinsics). To date, the best
available mechanism of achieving this result was to use a
Cassette-like pass rewriting every method and injecting
an overlay if necessary. However, this approach is somewhat
unsatisfying for two reasons:

1. It requires rewriting every function, which has non-trivial
   performance cost.
2. It is (currently) expensive because of the repeated calls to
   generated functions.
3. It confuses inference, because suddenly everything is one method.
   We have hooks to work around this, but support is incomplete.

It is also not clear that Cassette it is the best conceptual model,
because these methods *are* methods of the same generic function,
they just happen to only be applicable for a particular backend.

It is worth noting that this PR only gives the ability to keep
these tables of methods. It assigns no particular meaning to them
and the runtime (and regular inference) do not look at them.
They are designed as an implementation detail for external
compilers and similar tools.

This feature does not replace Cassette for the method-interception
use case in the absence of such a compiler, though it could in
the future form part of a solution (I'm hoping the AD work will
in due course lead to abstractions that would enable a "faster
Cassette" which may use part of these fetaures). As such,
I'm not sure we'll ever move this out of Experimental, but
until such a time that we have a better solution, I think this'll
be a useful feature for the GPU stack.

With all those disclaimers out of the way, here is a description
of the various parts of the current design that deserve
discussion:

 # Demo

```julia
julia> using Base.Experimental: @overlay, @MethodTable

julia> @MethodTable(mt)
 # 0 methods:

julia> mt
 # 0 methods:

julia> @overlay mt function sin(x::Float64)
           1
       end

julia> @overlay mt function cos(x::Float64)
           1
       end

julia> mt
 # 2 methods:
[1] cos(x::Float64) in Main at REPL[5]:1
[2] sin(x::Float64) in Main at REPL[4]:1

julia> Base._methods_by_ftype(Tuple{typeof(sin), Float64}, mt, 1, typemax(UInt))
1-element Vector{Any}:
 Core.MethodMatch(Tuple{typeof(sin), Float64}, svec(), sin(x::Float64) in Main at REPL[4]:1, true)

julia> Base._methods_by_ftype(Tuple{typeof(sin), Float64}, 1, typemax(UInt))
1-element Vector{Any}:
 Core.MethodMatch(Tuple{typeof(sin), Float64}, svec(Float64), sin(x::T) where T<:Union{Float32, Float64} in Base.Math at special/trig.jl:29, true)
```

 # The `@overlay` macro

The macro replaces the function name by an `Expr(:overlay, mt, name)`,
which then gets piped through to Method def. One particular design
aspect here is that I've stopped letting the 4-argument :method
Expr introduce new generic functions, reserving this functionality
entirely to the 2-argument :method Expr. We already started going
this way when we began omitting method names from the 4-argument
version. This PR re-uses that name field of the 4-argument version
to specify a method table instead.

 # Identity of methods

I think one of the biggest questions of this design is what happens
to the identity of methods. Until OpaqueClosure, all methods were uniquely
identified by their signatures, with the applicable method table
computed from the first argument of the signature. This is important
so that incremental compilation can properly merge method tables coming
from different .ji files. For these methods, that is of course not the
correct method table to use for these methods, so methods
that are not part of the internal method table will instead have a
backreference to the applicable method table.

 # Identity of method tables

Method tables are identified by the name of their binding in the
containing module. To ensure consistency of this mapping, these
MethodTables may only be constructed using the `@MethodTable(name)`
macro, which simultaneously establishes a const binding in
the declaring module.

TODO:
- [x] Core.Compiler interfaces to make this easy to use with
  external AbstractInterpreters
- [x] Serialization.jl support
- [x] Tests
- [x] Docs

Keno committed:
RFC: External Method Tables

This PR implements a way to keep tables of methods that are
not part of the internal method table, but still participate
in the special support we have for keeping tables of methods,
in particular unification through precompilation and efficient
lookup. The intended design use case is to allow for method overlay
tables for various non-CPU backends (e.g. GPU and TPU). These
backends would like to modify basic function like `sin` to
perform better on the device in question (or in the case of TPU
to define them over non-LLVM intrinsics). To date, the best
available mechanism of achieving this result was to use a
Cassette-like pass rewriting every method and injecting
an overlay if necessary. However, this approach is somewhat
unsatisfying for two reasons:

1. It requires rewriting every function, which has non-trivial
   performance cost.
2. It is (currently) expensive because of the repeated calls to
   generated functions.
3. It confuses inference, because suddenly everything is one method.
   We have hooks to work around this, but support is incomplete.

It is also not clear that Cassette it is the best conceptual model,
because these methods *are* methods of the same generic function,
they just happen to only be applicable for a particular backend.

It is worth noting that this PR only gives the ability to keep
these tables of methods. It assigns no particular meaning to them
and the runtime (and regular inference) do not look at them.
They are designed as an implementation detail for external
compilers and similar tools.

This feature does not replace Cassette for the method-interception
use case in the absence of such a compiler, though it could in
the future form part of a solution (I'm hoping the AD work will
in due course lead to abstractions that would enable a "faster
Cassette" which may use part of these fetaures). As such,
I'm not sure we'll ever move this out of Experimental, but
until such a time that we have a better solution, I think this'll
be a useful feature for the GPU stack.

With all those disclaimers out of the way, here is a description
of the various parts of the current design that deserve
discussion:

 # Demo

```julia
julia> using Base.Experimental: @overlay, @MethodTable

julia> @MethodTable(mt)
 # 0 methods:

julia> mt
 # 0 methods:

julia> @overlay mt function sin(x::Float64)
           1
       end

julia> @overlay mt function cos(x::Float64)
           1
       end

julia> mt
 # 2 methods:
[1] cos(x::Float64) in Main at REPL[5]:1
[2] sin(x::Float64) in Main at REPL[4]:1

julia> Base._methods_by_ftype(Tuple{typeof(sin), Float64}, mt, 1, typemax(UInt))
1-element Vector{Any}:
 Core.MethodMatch(Tuple{typeof(sin), Float64}, svec(), sin(x::Float64) in Main at REPL[4]:1, true)

julia> Base._methods_by_ftype(Tuple{typeof(sin), Float64}, 1, typemax(UInt))
1-element Vector{Any}:
 Core.MethodMatch(Tuple{typeof(sin), Float64}, svec(Float64), sin(x::T) where T<:Union{Float32, Float64} in Base.Math at special/trig.jl:29, true)
```

 # The `@overlay` macro

The macro replaces the function name by an `Expr(:overlay, mt, name)`,
which then gets piped through to Method def. One particular design
aspect here is that I've stopped letting the 4-argument :method
Expr introduce new generic functions, reserving this functionality
entirely to the 2-argument :method Expr. We already started going
this way when we began omitting method names from the 4-argument
version. This PR re-uses that name field of the 4-argument version
to specify a method table instead.

 # Identity of methods

I think one of the biggest questions of this design is what happens
to the identity of methods. Until OpaqueClosure, all methods were uniquely
identified by their signatures, with the applicable method table
computed from the first argument of the signature. This is important
so that incremental compilation can properly merge method tables coming
from different .ji files. For these methods, that is of course not the
correct method table to use for these methods, so methods
that are not part of the internal method table will instead have a
backreference to the applicable method table.

 # Identity of method tables

Method tables are identified by the name of their binding in the
containing module. To ensure consistency of this mapping, these
MethodTables may only be constructed using the `@MethodTable(name)`
macro, which simultaneously establishes a const binding in
the declaring module.

TODO:
- [] Core.Compiler interfaces to make this easy to use with
  external AbstractInterpreters
- [] Serialization.jl support
- [] Tests
- [] Docs

timholy commented:
More excitement for Revise! :laughing:

But this makes a lot of sense, so :+1:

maleadt committed:
Add an OverlayMethodTable type for use with method_table(::AbstractInterpreter).

maleadt commented:
`@overlay` can currently be used with functions in other modules without having to explicitly import/specify them (i.e. `@overlay sin` instead of `@overlay Base.sin`). That makes sense, as it can't be used to introduce new functions -- it should always refer to an existing one -- but maybe it should behave similar to extending functions? If so, where's the logic for that?

Precompilation also doesn't seem to work, looking into that.

Keno commented:
> Another design consideration: the current design doesn't allow adding methods to another module's overlay method table, or at least that won't survive precompilation. Maybe that's acceptable though?

This was supposed to work. I added extra code in serialization for it. I may have done it wrong though.

maleadt commented:
> This was supposed to work. I added extra code in serialization for it. I may have done it wrong though.

OK, good to know, I'll debug it then.

maleadt commented:
> I added extra code in serialization for it.

Where's that code? I only see the changes to refer to another module's method table when (de)serializing a method, but nothing to save or merge another module's method table? And just to be clear:

`Foo.jl`:
```julia
module Foo
  Base.Experimental.@MethodTable(mt)
end
```

`Bar.jl`:
```julia
module Bar
  using Foo
  Base.Experimental.@overlay Foo.mt sin(x::Int) = 1
end
```

```julia
using Foo, Bar
@show Foo.mt
```

```
Foo.mt = # 0 methods
```

maleadt committed:
Minimal docs and tests.

pabloferz commented on base/experimental.jl:
```diff
@@ -255,4 +256,37 @@ end
+    if !isexpr(def, [:function, :(=)]) || !isexpr(def.args[1], :call)
+        error("@overlay requires a function Expr")
+    end
+    name = def.args[1].args[1]
+    def.args[1].args[1] = Expr(:overlay, mt, def.args[1].args[1])
```
`name` here doesn't seem to be used.

maleadt commented:
@Keno Thoughts on https://github.com/JuliaLang/julia/pull/39697#issuecomment-790379610?
Also, what needs to happen for Serialization.jl? `serialize_any` seems to work fine for method tables; I assume you want it to maintain the same contract `@MethodTable` enforces (serialize the owning module and binding name, and assign it like that)? That seems like weird behaviour for Serialization's API though...

Keno commented on src/dump.c:
```diff
@@ -2461,7 +2492,7 @@ static jl_method_t *jl_recache_method(jl_method_t *m)
     assert(!m->is_for_opaque_closure);
     jl_datatype_t *sig = (jl_datatype_t*)m->sig;
-    jl_methtable_t *mt = jl_method_table_for((jl_value_t*)m->sig);
+    jl_methtable_t *mt = jl_method_get_table(m);
```
@maleadt this is where it's supposed to see the other method table and re-insert it there (together with the code above that fetches the reference to the mt).

maleadt replied:
The problem is that we never serialize (during precompilation) the overlay method when referring to a method table in another module, because we discover the methods to serialize by looking at what I assume are the internal method tables (`jl_type_type_mt` and `jl_nonfunction_mt`): https://github.com/JuliaLang/julia/blob/faa3d41f239adeef2692b2ad312a81f93fce8710/src/dump.c#L2204-L2219

Conversely, when the external method table is defined in the module that's being precompiled, we just serialize that object, and the fallback serialization that recurses into the method table's fields ends up serializing the overlay method.

maleadt committed:
Collect lambdas from external method tables during deserialization.

maleadt committed:
Refuse serializing methods that are part of external method tables.

Serialization.jl is only for serializing plain data, and
does not implement the necessary linking functionality.

maleadt committed:
Move external_mt field to the uninitialized part of jl_method_t.

Otherwise the isdefined tfunc breaks.

maleadt commented:
Ah, I forgot we hadn't merged this yet. @Keno, do you want to fold this in your upcoming compiler plugin work, or should we merge this first?

Keno commented:
I'm happy merging this as is and revising it later if necessary.
